main/src/ca/uwaterloo/flix/language/phase/Namer.scala

```diff
@@ -71,214 +71,229 @@ object Namer extends Phase[WeededAst.Program, NamedAst.Root] {
   /**
     * Performs naming on the given declaration `decl0` in the given namespace `ns0` under the given (partial) program `prog0`.
     */
-  private def visitDecl(decl0: WeededAst.Declaration, ns0: Name.NName, uenv0: UseEnv, prog0: NamedAst.Root)(implicit flix: Flix): Validation[NamedAst.Root, NameError] = decl0 match {
-    /*
-     * Namespace.
-     */
-    case WeededAst.Declaration.Namespace(ns, uses, decls, loc) =>
-      mergeUseEnvs(uses, uenv0) flatMap {
-        newEnv =>
-          Validation.fold(decls, prog0) {
-            case (pacc, decl) =>
-              val namespace = Name.NName(ns.sp1, ns0.idents ::: ns.idents, ns.sp2)
-              visitDecl(decl, namespace, newEnv, pacc)
-          }
-      }
+  private def visitDecl(decl0: WeededAst.Declaration, ns0: Name.NName, uenv0: UseEnv, prog0: NamedAst.Root)(implicit flix: Flix): Validation[NamedAst.Root, NameError] = {
 
-    case decl@WeededAst.Declaration.Class(doc, mod, ident, tparam, sigs, loc) =>
-      // Check if the class already exists.
-      val classes = prog0.classes.getOrElse(ns0, Map.empty)
-      val sigNs = Name.extendNName(ns0, ident)
-      val defsAndSigs0 = prog0.defsAndSigs.getOrElse(sigNs, Map.empty)
-      classes.get(ident.name) match {
-        case None =>
-          // Case 1: The class does not already exist. Update it.
-          visitClass(decl, uenv0, Map.empty, ns0) flatMap {
-            case clazz@NamedAst.Class(_, _, _, _, sigs, _) =>
-              // add each signature to the namespace
-              val defsAndSigsVal = Validation.fold(sigs, defsAndSigs0) {
-                case (defsAndSigs, sig) => defsAndSigs.get(sig.sym.name) match {
-                  case Some(otherSig) =>
-                    val name = sig.sym.name
-                    val loc1 = sig.loc
-                    val loc2 = otherSig.loc
-                    Failure(LazyList(
-                      // NB: We report an error at both source locations.
-                      NameError.DuplicateDefOrSig(name, loc1, loc2),
-                      NameError.DuplicateDefOrSig(name, loc2, loc1)
-                    ))
-                  case None => (defsAndSigs + (sig.sym.name -> sig)).toSuccess
+    /**
+      * Creates a pair of errors reporting a duplicate type declaration at each location.
+      */
+    def mkDuplicateTypeOrClassPair[T](name: String, loc1: SourceLocation, loc2: SourceLocation): Validation.Failure[T, NameError] = {
```

Comment:
Move this helper function somewhere near the bottom.

Also, can we come up with a better name?

Reply:
I changed the name in my push, not sure it's much better now.